single_kernel_mongo/managers/backups.py

```diff
@@ -0,0 +1,570 @@
+from tenacity import (
+    Retrying,
+    before_log,
+    retry,
+    retry_if_exception_type,
+    retry_if_not_exception_type,
+    stop_after_attempt,
+    wait_fixed,
+)
+
+from single_kernel_mongo.config.literals import MongoPorts, Substrates
+from single_kernel_mongo.config.models import Role
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.exceptions import (
+    BackupError,
+    InvalidArgumentForActionError,
+    InvalidPBMStatusError,
+    ListBackupError,
+    PBMBusyError,
+    RestoreError,
+    ResyncError,
+    SetPBMConfigError,
+    WorkloadExecError,
+)
+from single_kernel_mongo.managers.config import BackupConfigManager
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.workload import get_pbm_workload_for_substrate
+from single_kernel_mongo.workload.backup_workload import PBMWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm  # pragma: nocover
+
+BACKUP_RESTORE_MAX_ATTEMPTS = 10
+BACKUP_RESTORE_ATTEMPT_COOLDOWN = 15
+REMAPPING_PATTERN = r"\ABackup doesn't match current cluster topology - it has different replica set names. Extra shards in the backup will cause this, for a simple example. The extra/unknown replica set names found in the backup are: ([\w\d\-,\s]+)([.] Backup has no data for the config server or sole replicaset)?\Z"
+
+S3_PBM_OPTION_MAP = {
+    "region": "storage.s3.region",
+    "bucket": "storage.s3.bucket",
+    "path": "storage.s3.prefix",
+    "access-key": "storage.s3.credentials.access-key-id",
+    "secret-key": "storage.s3.credentials.secret-access-key",
+    "endpoint": "storage.s3.endpointUrl",
+    "storage-class": "storage.s3.storageClass",
+}
+logger = logging.getLogger(__name__)
+
+
+def _backup_restore_retry_before_sleep(retry_state) -> None:
+    logger.error(
+        f"Attempt {retry_state.attempt_number} failed. {BACKUP_RESTORE_MAX_ATTEMPTS - retry_state.attempt_number} attempts left."
+        f"Retrying after {BACKUP_RESTORE_ATTEMPT_COOLDOWN} seconds."
+    )
+
+
+class BackupManager(Object, BackupConfigManager):
+    """Manager for the S3 integrator and backups."""
+
+    def __init__(
+        self,
+        charm: AbstractMongoCharm,
+        role: Role,
+        substrate: Substrates,
+        state: CharmState,
+        container: Container | None,
+    ) -> None:
+        super().__init__(parent=charm, key="backup")
+        super(Object, self).__init__(
+            role=role,
+            substrate=substrate,
+            config=charm.parsed_config,
+            state=state,
+            container=container,
+        )
+        self.charm = charm
+        self.workload: PBMWorkload = get_pbm_workload_for_substrate(substrate)(
+            role=role, container=container
+        )
+        self.state = state
+
+    @cached_property
+    def environment(self) -> dict[str, str]:
+        """The environment used to run PBM commands."""
```

Comment:
can we add some context here so it is more maintable 
```suggestion
        """
        The environment used to run PBM commands.
        
        In the case of PBM this is the URI, we never pass the URI via arguments to the PBM command for security reasons.
        """
```

Reply:
Sure I'll update the documentation.